tss-esapi/src/traits.rs

```diff
@@ -51,65 +38,151 @@ pub trait UnMarshall: Sized {
-        marshalled_data: &[u8],
-        offset: &mut std::os::raw::c_ulong,
-    ) -> Result<Self> {
-        let mut dest = 0_u32;
+/// Macro used to implement Marshall and Unmarshall for types that
+/// can be converted from native to TSS i.e. it cannot fail and are
+/// passed to MUAPI by value(i.e. pointer).
+macro_rules! impl_mu_simple {
+    ($native_type:ident, $tss_type:ident, $tss_mu_type:ident) => {
+        $crate::traits::impl_marshall_trait!($native_type, $tss_type, $tss_mu_type, into(),);
+        $crate::traits::impl_unmarshall_trait!(
+            $native_type,
+            $tss_type,
+            $tss_mu_type,
+            $native_type::try_from
+        );
+    };
+    ($native_type:ident, $tss_type:ident) => {
+        $crate::traits::impl_mu_simple!($native_type, $tss_type, $tss_type);
+    };
+}
 
-        ReturnCode::ensure_success(
-            unsafe {
-                crate::tss2_esys::Tss2_MU_UINT32_Unmarshal(
-                    marshalled_data.as_ptr(),
-                    marshalled_data.len().try_into().map_err(|e| {
-                        error!("Failed to convert length of marshalled data: {}", e);
-                        Error::local_error(WrapperErrorKind::InvalidParam)
-                    })?,
-                    offset,
-                    &mut dest,
-                )
-            },
-            |ret| error!("Failed to unmarshal SensitiveCreate: {}", ret),
-        )?;
+/// Macro used to implement Marshall and Unmarshall for types that
+/// can be converted from native to TSS without failing and are
+/// passed to MUAPI by reference(i.e. pointer).
+macro_rules! impl_mu_standard {
+    ($native_type:ident, $tss_type:ident, $tss_mu_type:ident) => {
+        $crate::traits::impl_marshall_trait!($native_type, $tss_type, $tss_mu_type, into(), &);
+        $crate::traits::impl_unmarshall_trait!(
+            $native_type,
+            $tss_type,
+            $tss_mu_type,
+            $native_type::try_from
+        );
+    };
+    ($native_type:ident, $tss_type:ident) => {
+        $crate::traits::impl_mu_standard!($native_type, $tss_type, $tss_type);
+    };
+}
 
-        Ok(dest)
-    }
+/// Macro used to implement Marshall and Unmarshall for types that
+/// can be converted from native to TSS with the possibility of failing
+/// and are passed to MUAPI by reference(i.e. pointer).
+macro_rules! impl_mu_complex {
+    ($native_type:ident, $tss_type:ident, $tss_mu_type:ident) => {
+        $crate::traits::impl_marshall_trait!($native_type, $tss_type, $tss_mu_type, try_into()?, &);
+        $crate::traits::impl_unmarshall_trait!(
+            $native_type,
+            $tss_type,
+            $tss_mu_type,
+            $native_type::try_from
+        );
+    };
+    ($native_type:ident, $tss_type:ident) => {
+        $crate::traits::impl_mu_complex!($native_type, $tss_type, $tss_type);
+    };
 }
+
+// Make the macros usable outside of the module.
+pub(crate) use impl_marshall_trait;
```

Comment:
Is this actually imported somewhere else?

Reply:
No, but I thought it could be useful if you for some reason only wanted to implement one of the traits or there was a case where the implementation did not fit the other macros.